mlflow/java/client/src/main/java/org/mlflow/utils/CliCommand.java

```diff
@@ -0,0 +1,132 @@
+package org.mlflow.utils;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import org.apache.commons.io.IOUtils;
+import org.mlflow.tracking.MlflowClientException;
+import org.mlflow.tracking.creds.MlflowHostCreds;
+import org.mlflow.tracking.creds.MlflowHostCredsProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Runs MLflow CLI commands
+ *
+ * TODO(andy): CliBasedArtifactRepository was the progenitor to this function. It contains
```

Comment:
yeah i think we should discuss this -- the python `ArtifactRepository` is somewhat run-based but not explicitly (it has an artifact_root but it does not explicitly say that is based on a run). I think the java `CliArtifactRepository` has `runId` for convenience, not because it's meant to be run-based. it'll take some work to re-work `CliArtifactRepository` (which seems more ideal than having to duplicate the functionality here) so we should discuss when you get back.

Reply:
Agreed. I'm happy to discuss this and if we can come up with a way to use the ArtifactRepository, I'd like to do it.

---
Below is some of my thoughts on it.

I originally intended to implement it using the `CliBasedArtifactRepository`. The problem I ran into is that the `runId` is part of the class's constructor. 

In order to refactor `CliBasedArtifactRepository` I would need to have a way to construct a `CliBasedArtifactRepository` without the runId, we have several options.

1. Pass in nulls into the existing constructor. I don't recommend this because it means that that while `downloadArtifactUri` may work, the other functions will only generate an error. 
2. Change the base `ArtifactRepository` interface to include `runId` as a parameter which would be a backwards breaking change. I don't recommend this because it will break anybody who is using the methods.
3. Extend `CliBasedArtifactRepository` and throw `UnsupportedOperationException`. I could but it turns everything into a runtime exception and will surprise users if they try to use it.
4. Create a new interface and implementing class. I don't recommend this either because it's not the `ArtifactRepository` interface at this point.